modules/ximgproc/src/fld.cpp

```diff
@@ -0,0 +1,794 @@
+                    j--;
+                    break;
+                }
+            }
+            pe = pt;
+            l_points.push_back(pt);
+        }
+        fitLine( Mat(l_points), line, DIST_L2, 0, 0.01, 0.01);
+        a[0] = line[2];
+        a[1] = line[3];
+        b[0] = line[2] + line[0];
+        b[1] = line[3] + line[1];
+
+        p1 = Mat(3, 1, CV_64FC1, a).clone();
+        p2 = Mat(3, 1, CV_64FC1, b).clone();
+
+        l = p1.cross(p2);
+
+        Point2f e1, e2;
+        e1.x = (float)ps.x;
+        e1.y = (float)ps.y;
+        e2.x = (float)pe.x;
+        e2.y = (float)pe.y;
+
+        incidentPoint(l, e1);
+        incidentPoint(l, e2);
+        seg.x1 = e1.x;
+        seg.y1 = e1.y;
+        seg.x2 = e2.x;
+        seg.y2 = e2.y;
+
+        segments.push_back(seg);
+        i = i + j;
+    }
+}
+
+void FastLineDetectorImpl::pointInboardTest(const Mat& src, Point2i& pt)
+{
+    pt.x = pt.x <= 5 ? 5 : pt.x >= src.cols - 5 ? src.cols - 5 : pt.x;
+    pt.y = pt.y <= 5 ? 5 : pt.y >= src.rows - 5 ? src.rows - 5 : pt.y;
+}
+
+bool FastLineDetectorImpl::getPointChain( const Mat& img, Point pt, Point& chained_pt, float& direction, int step )
+{
+    int ri, ci;
+    int indices[8][2]={ {1,1}, {1,0}, {1,-1}, {0,-1}, {-1,-1},{-1,0}, {-1,1}, {0,1} };
+
+    float min_dir_diff = 7.0f;
+    Point consistent_pt;
+    int consistent_direction = 0;
+    for ( int i = 0; i < 8; i++ )
+    {
+        ci = pt.x + indices[i][1];
+        ri = pt.y + indices[i][0];
+
+        if ( ri < 0 || ri == img.rows || ci < 0 || ci == img.cols )
+            continue;
+
+        if ( img.at<unsigned char>(ri, ci) == 0 )
+            continue;
+
+        if(step == 0)
+        {
+            chained_pt.x = ci;
+            chained_pt.y = ri;
+            // direction = (float)i;
+            direction = i > 4 ? (float)(i - 8) : (float)i;
+            return true;
+        }
+        else
+        {
+            float curr_dir = i > 4 ? (float)(i - 8) : (float)i;
+            float dir_diff = abs(curr_dir - direction);
+            dir_diff = dir_diff > 4.0f ? 8.0f - dir_diff : dir_diff;
+            if(dir_diff <= min_dir_diff)
+            {
+                min_dir_diff = dir_diff;
+                consistent_pt.x = ci;
+                consistent_pt.y = ri;
+                consistent_direction = i > 4 ? i - 8 : i;
+            }
+        }
+    }
+    if(min_dir_diff < 2.0f)
+    {
+        chained_pt.x = consistent_pt.x;
+        chained_pt.y = consistent_pt.y;
+        direction = (direction * (float)step + (float)consistent_direction)
+            / (float)(step + 1);
+        return true;
+    }
+    return false;
+}
+
+void FastLineDetectorImpl::lineDetection(const Mat& src, std::vector<SEGMENT>& segments_all)
+{
+    int r, c;
+    imageheight=src.rows; imagewidth=src.cols;
+
+    std::vector<Point2i> points;
+    std::vector<SEGMENT> segments, segments_tmp;
+    Mat canny;
+    Canny(src, canny, 50, 50, 3);
```

Comment:
Canny result depends on image quality (contrast). There is no universal parameters to get good edge maps in all cases.

Probably this method may accept "edge" map directly instead of original image.
Or Canny parameters may be exposed via public interface.